crates/symbolicator-service/src/download/index.rs

```diff
@@ -0,0 +1,252 @@
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_index_segment(
+            downloader,
+            source,
+            self.segment_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+impl CacheItemRequest for FetchSymstoreIndex {
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_full_index(
+            self.segment_cache.clone(),
+            downloader,
+            source,
+            self.scope.clone(),
+            self.last_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SymstoreIndexService {
+    cache: Arc<Cacher<FetchSymstoreIndex>>,
+    segment_cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    downloader: Arc<DownloadService>,
+    last_id_cache: moka::future::Cache<(Scope, SourceId), CacheContents<u32>>,
+}
+
+impl SymstoreIndexService {
+    pub fn new(
+        cache: Cache,
+        shared_cache: SharedCacheRef,
+        downloader: Arc<DownloadService>,
+    ) -> Self {
+        let last_id_cache = moka::future::Cache::builder()
+            .max_capacity(100)
+            .name("last_id")
+            .time_to_live(Duration::from_secs(24 * 60 * 60))
+            .build();
+
+        Self {
+            cache: Arc::new(Cacher::new(cache.clone(), shared_cache.clone())),
+            segment_cache: Arc::new(Cacher::new(cache, shared_cache)),
+            downloader,
+            last_id_cache,
+        }
+    }
+
+    async fn get_symstore_last_id(
+        &self,
+        scope: Scope,
+        source: Arc<HttpSourceConfig>,
+    ) -> CacheContents<u32> {
+        self.last_id_cache
+            .get_with((scope, source.id.clone()), async {
+                let temp_file = NamedTempFile::new()?;
+                let loc = "000Admin/lastid.txt";
```

Comment:
I think it would be nice if this was a constant and we documentation on how this index is layed out.